amethyst_network/src/simulation/resource.rs

```diff
@@ -0,0 +1,192 @@
+use super::{
+    client::Client,
+    message::Message,
+    requirements::{DeliveryRequirement, UrgencyRequirement},
+    transport::socket::Socket,
+};
+use std::{collections::VecDeque, net::SocketAddr};
+
+/// Resource serving as the owner of the underlying socket and the queue of messages to be sent.
+pub struct NetworkSimulationResource<S: Socket> {
+    pub(crate) socket: Option<S>,
+    is_server: bool,
+    server_addr: Option<SocketAddr>,
+    clients: Vec<Client>,
+    pub(crate) messages: VecDeque<Message>,
+}
+
+impl<S: Socket> NetworkSimulationResource<S> {
+    /// Create a new `NetworkSimulationResource` as a client
+    pub fn new_client(server_addr: SocketAddr) -> Self {
+        Self {
+            socket: None,
+            server_addr: Some(server_addr),
+            clients: Vec::new(),
+            messages: VecDeque::new(),
+            is_server: false,
+        }
+    }
+
+    /// Create a new `NetworkSimulationResource` as a server
+    pub fn new_server() -> Self {
+        Self {
+            socket: None,
+            server_addr: None,
+            clients: Vec::new(),
+            messages: VecDeque::new(),
+            is_server: true,
+        }
+    }
+
+    /// Add a number of trusted clients to the `NetworkSimulationResource` for use as a server
+    pub fn with_trusted_clients(mut self, clients: &[SocketAddr]) -> Self {
```

Comment:
This might be unnecessary in the future. I would like to move this functionality out of our library. Let the user decide what to do with connections. Ignore, accept, deny them. Because in most cases we would never know what our client IP's would be.

Reply:
Well that's an interesting question. For example, if you had a matchmaking service, before the server code starts running, you would probably know all of the users in the lobby.

Reply:
Right, but people might as well leave and join during the game after the lobby. So if we would use all IP's from the lobby, we can't have people joining during the game.

Reply:
Yeah that's fair. I'll put more thought into this. It might even be worth lifting this concept up into a higher abstraction level.